allennlp/training/callback_trainer.py

```diff
@@ -146,6 +148,18 @@ def __init__(
         # For capturing errors that occur during the train loop.
         self.exception: Optional[Exception] = None
 
+        # Using `DistributedDataParallel`(ddp) brings in a quirk wrt AllenNLP's `Model` interface and its
+        # usage. A `Model` object is wrapped by `ddp`, but assigning the wrapped model to `self.model`
+        # will break the usages such as `Model.get_regularization_penalty`, `Model.get_metrics`, etc.
```

Comment:
> Have you and @DeNeutoy talked much about how metrics will work across processes? 

Metrics aggregation across processes is handled in `training_util.get_metrics` already which has been done in #3372 . So that part should just be fine for now I guess.

> The regularization penalty also seems like it could be an issue...

I'm not sure if that would be an issue as the penalty is computed from the `named_parameters` of the model. Since the gradients are synced across workers before the optimization step, this might just work well as is. Please correct me if I'm wrong. 

The reason I added the said comment is primarily to highlight the inconsistencies that may arise if we wrap `allennlp.model.Model` with `DistributedDataParallel` and use the same object throughout the trainer. Wrapping is necessary and it would break the `Model` specific interface methods such as `get_regularization_penalty`, etc. Hence the wrapped reference with a different name (`self._pytorch_model`) will only be used for pytorch specific ops such as model forward, etc. and the usages will only be through the actual `Model` instance.

Reply:
I think this makes sense. Thanks for explaining!